pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -173,41 +197,99 @@ func (c *Cluster) Start() error {
+			errorMessages = append(errorMessages, fmt.Sprintf("empty volumes for node %s", n.Name))
+			continue
+		}
+		// create the job that prepares osds on the node
+		job, err := c.makeJob(n.Name, devicesToUse[n.Name], n.Selection, n.Resources, storeConfig, metadataDevice, n.Location)
+		if err != nil {
+			message := fmt.Sprintf("failed to create prepare job node %s: %v", n.Name, err)
+			logger.Info(message)
+			status := OrchestrationStatus{Status: OrchestrationStatusCompleted, Message: message}
+			if err := UpdateOrchestrationStatusMap(c.context.Clientset, c.Namespace, n.Name, status); err != nil {
+				errorMessages = append(errorMessages, message)
+				continue
+			}
+		}
+		_, err = c.context.Clientset.Batch().Jobs(c.Namespace).Create(job)
 		if err != nil {
 			if !errors.IsAlreadyExists(err) {
-				// we failed to create the replica set, update the orchestration status for this node
-				message := fmt.Sprintf("failed to create osd replica set for node %s. %+v", n.Name, err)
+				// we failed to create job, update the orchestration status for this node
+				message := fmt.Sprintf("failed to create osd prepare job for node %s. %+v", n.Name, err)
 				c.handleOrchestrationFailure(*n, message, &errorMessages)
+				err = discover.FreeDevices(c.context, n.Name, clusterName)
+				if err != nil {
+					logger.Warningf("failed to free devices: %s", err)
+				}
 				continue
 			} else {
-				// TODO: if the replica set already exists, we may need to edit the pod template spec, for example if device filter has changed
-				message := fmt.Sprintf("osd replica set already exists for node %s", n.Name)
+				// TODO: if the job already exists, we may need to edit the pod template spec, for example if device filter has changed
+				message := fmt.Sprintf("failed to set orchestration status for node %s, status: %+v: %+v", n.Name, status, err)
 				logger.Info(message)
 				status := OrchestrationStatus{Status: OrchestrationStatusCompleted, Message: message}
 				if err := UpdateOrchestrationStatusMap(c.context.Clientset, c.Namespace, n.Name, status); err != nil {
-					errorMessages = append(errorMessages, fmt.Sprintf("failed to set orchestration status for node %s, status: %+v: %+v", n.Name, status, err))
+					errorMessages = append(errorMessages, message)
 					continue
 				}
 			}
 		} else {
-			logger.Infof("osd replica set started for node %s", n.Name)
+			logger.Infof("osd prepare job started for node %s", n.Name)
 		}
+	}
+	for i := range c.Storage.Nodes {
+		// fully resolve the storage config and resources for this node
+		n := c.resolveNode(c.Storage.Nodes[i])
+		storeConfig := config.ToStoreConfig(n.Config)
+		metadataDevice := config.MetadataDevice(n.Config)
 
 		// wait for the current node's orchestration to be completed
-		if err := c.waitForCompletion(n.Name); err != nil {
-			errorMessages = append(errorMessages, err.Error())
+		if status, err := c.waitForCompletion(n.Name); err != nil {
```

Comment:
if we have one bad node where the job never completes or takes a long time, this could block many nodes from starting their provisioned osds, right?

Reply:
will this call timeout after too long?

Reply:
I feel we might need a comprehensive node and pod health monitor. In this case if there is a bad node or provisioning pod goes wrong, it is more beneficial to remove the bad node from watch than timeout. If this is the way to go, I'll start it in a separate proposal.